### PR TITLE
feat: pull languages cmd

### PR DESCRIPTION
**Pull request by alvarosabu**
Hi, this is the new implementation of the `pull-components` command for the renovated version of the CLI.

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Other (please describe):

## How to test this PR

1. Clone the repo and pull the branch
2. Add several languages to a Storyblok Space or use the default test one `295017` or `295018`
3. Run `pnpm install`
4. Run `pnpm run dev pull-languages --space <SPACE_ID>`

### Testing checklist

- [x] pull languages with a space with languages configured.
  - [x] should create a `languages.295017.json` file on the root containing an object 
  ```json
  {
    "languages": [
      {
        "code": "de",
        "name": "German"
      },
      {
        "code": "ja",
        "name": "Japanese"
      },
      {
        "code": "pl",
        "name": "Polish"
      }
    ]
  }
  ```
  - [x] file should contain `"default_lang_name": "en",` if configured on the space
  - [x] should prompt an error if flag space is missing `Please provide the space as argument --space YOUR_SPACE_ID.`
  - [x] should prompt an API error if the space ID is incorrect
  - [x] should prompt an unauthorized error if the user is not logged in before operating this command

### CODE REVIEW

File: src/commands/pull-languages/actions.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+export const pullLanguages = async (space: string, token: string, region: string): Promise<SpaceInternationalizationOptions | undefined> => {
+  try {
+    const response = await ofetch(`https://${regionsDomain[region]}/v1/spaces/${space}`, {
+      headers: {
+        Authorization: token,
+      },
+    })
+    return {
```
Comment (edodusi) [inline src/commands/pull-languages/actions.ts:24]: I think this return statement would be better placed after the `catch` block, as it's not part of the code that can throw and also for better readability - you can even put it inside a `finally` block

File: src/commands/pull-languages/actions.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      await mkdir(resolvedPath, { recursive: true })
+    }
+
+    return await writeFile(filePath, data, {
+      mode: 0o600,
+    })
+  }
+  catch (error) {
+    handleFileSystemError('write', error as Error)
+  }
```
Comment (edodusi) [inline src/commands/pull-languages/actions.ts:64]: this code could also be refactored to follow the try...catch flow, like this for example

```
import { writeFile, mkdir, access, constants } from 'fs/promises'
import { resolve, join } from 'path'
import { SpaceInternationalizationOptions } from 'actual-module-name' // Replace with the actual module name
import { handleFileSystemError } from 'actual-module-name' // Replace with the actual module name

export const saveLanguagesToFile = async (space: string, internationalizationOptions: SpaceInternationalizationOptions, path?: string) => {
  const data = JSON.stringify(internationalizationOptions, null, 2)
  const filename = `languages.${space}.json`
  const resolvedPath = path ? resolve(process.cwd(), path) : process.cwd()
  const filePath = join(resolvedPath, filename)

  try {
    await access(resolvedPath, constants.F_OK)
  } catch {
    try {
      await mkdir(resolvedPath, { recursive: true })
    } catch (mkdirError) {
      handleFileSystemError('mkdir', mkdirError as Error)
    }
  }

  try {
    await writeFile(filePath, data, { mode: 0o600 })
  } catch (error) {
    handleFileSystemError('write', error as Error)
  }
}
 ```

File: src/commands/pull-languages/index.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+    if (!space) {
+      handleError(new CommandError(`Please provide the space as argument --space YOUR_SPACE_ID.`), verbose)
+      return
+    }
+
+    try {
+      const internationalization = await pullLanguages(space, state.password, state.region)
+
+      if (!internationalization || internationalization.languages.length === 0) {
```
Comment (edodusi) [inline src/commands/pull-languages/index.ts]: can `languages` be undefined? in this case I will be more resilient with

`if (!internationalization || !internationalization.languages || internationalization.languages.length === 0) {`

**Review by edodusi — CHANGES_REQUESTED** [review]
Adding some code comments, proceeding to manual testing!

**Comment by edodusi** [comment]
@alvarosabu I manually tested and checked everything 🚀 

One thing, the steps you provided mention that I should have `"default_lang_name": "en",` in the json file, while I have `"default_lang_name": "english",`, is this correct?

**Comment by alvarosabu** [comment]
> @alvarosabu I manually tested and checked everything 🚀
> 
> One thing, the steps you provided mention that I should have `"default_lang_name": "en",` in the json file, while I have `"default_lang_name": "english",`, is this correct?

Awesome, thanks for testing it.

Thats weird. My JSON looks like this:

```
{
  "default_lang_name": "en",
  "languages": [
    {
      "code": "de",
      "name": "German"
    },
    {
      "code": "ja",
      "name": "Japanese"
    },
    {
      "code": "pl",
      "name": "Polish"
    }
  ]
}
```

@edodusi Which space are you using? Send me via Slack

File: src/commands/pull-languages/actions.ts (hunk 1 above)
Comment (alvarosabu) [reply src/commands/pull-languages/actions.ts:24]: Hi, @edodusi the return should remain inside of the try-catch.

If you add it after, the code can't find the name `response` 

Screenshot 2024-11-05 at 09 14 04


finally, it is also not a safe option in this case, check https://eslint.org/docs/latest/rules/no-unsafe-finally

Screenshot 2024-11-05 at 09 13 29

Comment (edodusi) [reply src/commands/pull-languages/actions.ts:24]: right, that makes sense

**Review by edodusi — APPROVED** [review]
👍

**Pull request merged**